### PR TITLE
Add `validate_data` for single table

### LINKED ISSUE #1518: Add validate_data method to Metadata
### Problem Description

_As a user, it would be convenient to validate my metadata against a dataset without having to make a synthesizer._

Currently the code for validating data is in the [synthesizer](https://github.com/sdv-dev/SDV/blob/c4a52305da59ed589f9d044a696a9791ee2bd1cc/sdv/single_table/base.py#L218). We can move any code that is specific to metadata validation out into the metadata class itself.

### Expected behavior

- Add `validate_with_data` method to metadata.
- Parameters:
    - `data`:
        - For single table, a pandas.DataFrame with the data
        - For multi table, a dictionary mapping each table name to a pandas.DataFrame

### Additional context

- The synthesizer validation has some code that is specific to certain synthesizers, as well as constraint validation. This method only needs to validate things that can be found in the metadata alone (ie. `_validate_keys_dont_have_missing_values`, `_validate_key_values_are_unique`, `_validate_metadata_matches_data`, `_valdiate_column`).

### PR DESCRIPTION
**Pull request by fealho**
Resolve #1518 for single-table.

### CODE REVIEW

**Review by frances-h — APPROVED** [review]
LGTM 👍

**Comment by fealho** [comment]
@amontanez24 Wouldn't it be better to call this `validate_data`, instead of `validate_with_data`? You are validating the data matches the metadata, so the `with` doesn't seem necessary (it's also harder to remember, `validate_data` just seems a little more intuitive).

**Comment by amontanez24** [comment]
> @amontanez24 @npatki Wouldn't it be better to call this `validate_data`, instead of `validate_with_data`? You are validating the data matches the metadata, so the `with` doesn't seem necessary (it's also harder to remember, `validate_data` just seems a little more intuitive).

I'm not against it

**Comment by npatki** [comment]
In this method, we _won't_ validate the metadata format right? ie. the checks in the existing `metadata.validate()` and this method are completely different?

If so, then `validate_data()` makes sense to me, as we are only validating that the data matches the metadata. If we all agree, I suggest we update the issue to match!

File: sdv/metadata/single_table.py [hunk 1]
```diff
@@ -471,14 +478,173 @@ def validate(self):
...
+            invalid_values = self._format_invalid_values_string(invalid_values)
+            return [f"Invalid values found for {sdtype} column '{column.name}': {invalid_values}."]
+
+        return []
+
+    def _validate_constraints(self, data):
+        """Validate that the data satisfies the constraints."""
+        errors = []
+        try:
+            self._data_processor._fit_constraints(data)
```
Comment (amontanez24) [inline sdv/metadata/single_table.py:609]: the data_processor doesn't exist in the metadata. This shouldn't work. I think this got accidentally copied over

File: sdv/metadata/single_table.py [hunk 2]
```diff
@@ -471,14 +478,173 @@ def validate(self):
...
+            if pd.isna(data[key]).any():
+                errors.append(f"Key column '{key}' contains missing values.")
+
+        return errors
+
+    @staticmethod
+    def _format_invalid_values_string(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
```
Comment (amontanez24) [inline sdv/metadata/single_table.py:549]: A similar thing happens here:
https://github.com/sdv-dev/SDV/blob/6480f1f066ef302d718d147887adda3b588a495f/sdv/constraints/base.py#L204-L210

Maybe we can generalize this function into something that takes a dataframe and number of rows, and returns the string with the top x rows + y more as a string. Then we can move it into utils

**Review by amontanez24 — CHANGES_REQUESTED** [review]
This looks really good! Couple of small comments

File: sdv/metadata/single_table.py (hunk 1 above)
Comment (fealho) [reply sdv/metadata/single_table.py:609]: Yeah, this method doesn't get called, I copied it by accident. Good catch!

**Review by amontanez24 — APPROVED** [review]
Thanks for addressing the changes! :shipit:

**Pull request merged**